scheduler/src/cook/cors.clj

```diff
@@ -0,0 +1,59 @@
+(ns cook.cors)
+
+(defn preflight?
+  [req]
+  (= :options (:request-method req)))
+
+(defn same-origin?
+  "Returns true if the request is from the same origin as the provided origin header"
+  [{:keys [headers scheme]}]
+  (let [{:strs [host origin]} headers]
+    (when (and host origin scheme)
+      (= origin (str (name scheme) "://" host)))))
+
+(defn request-allowed?
+  "Returns true if the request is either from the same origin or matches a pattern in cors-origins"
+  [req cors-origins]
+  (or (same-origin? req)
+      (let [origin (get-in req [:headers "origin"])]
+        (some #(re-matches % origin) cors-origins))))
+
+(defn wrap-preflight
+  "Middleware for supporting CORS preflight requests"
+  [handler cors-origins]
+  (fn preflight-handler [{:keys [headers request-method] :as req}]
+    (let [{:strs [origin]} headers]
+      (if (and (preflight? req) origin)
```

Comment:
Can we flip the arguments so that `origin` is checked first?

Reply:
Should we respond with `403` `No origin provided` when origin is absent, it will simplify debugging for the client?

Reply:
I think it's possible for OPTIONS to be used for something other than CORS preflight, so I pass the request through to the underlying handlers.